src/app/state/rest/useApiCache.ts

```diff
@@ -0,0 +1,32 @@
+import { useCallback, useMemo, useReducer } from "react"
+
+export type ApiCache = {
+  cache: Record<string, any>
+  setItem: (key: string, value: any) => void
+  clear: () => void
+}
+
+type State = Record<string, any>
+type Action =
+  | { type: "SET_ITEM", key: string, value: any }
+  | { type: "CLEAR" }
+
+const reducer = (state: State, action: Action) => {
+  switch(action.type) {
+    case "SET_ITEM":
+      return { ...state, [action.key]: action.value }
+    case "CLEAR":
+      return {}
+  }
+}
+
+export const useApiCache = (): ApiCache => {
+  const [ cache, dispatch ] = useReducer(reducer, {})
+  // @ts-ignore
+  window.apiCache = cache
+
+  const setItem = useCallback((key: string, value: any) => dispatch({ type: "SET_ITEM", key, value }), [ dispatch ])
+  const clear = useCallback(() => dispatch({ type: "CLEAR" }), [ dispatch ])
```

Comment:
Any pending requests should be cancelled before clearing. Otherwise a situation might occur where an old pending request populates a cleared cache.

Reply:
I honestly think that would bring a lot of extra complexity for such a small edge-case.
Cache clear doesn't happen that often: only when the user submits a form, or hits a delete button.

To make the problem smaller we could give useApiRequest a separate cache-store as parameter. (think of them as groups).

Something like this (pseudo-code)
```
const caseApiStore = useApiStore()
export const useCases = () => {
  useApiRequest<APIListResponse<API.Case>>({
    url: slashSandwich([process.env.REACT_APP_GATEWAY, "cases"]),
    cache: caseApiStore // <- share store between hooks 
  })

export const useCase = () =>
  useApiRequest<APIListResponse<API.Case>>({
    url: slashSandwich([process.env.REACT_APP_GATEWAY, "case", id]),
    cache: caseApiStore // <- share store between hooks 
  })


```

Reply:
Do not agree that this is "a small edge-case". It's a big part of the problem that an async cache is supposed to solve.

Reply:
Could you explain why? :-)

Don't think a lot of GET requests within the same group (!) would still be pending when a POST in the same group happens and is also finished earlier.

The thing is:
- cancelling requests, while also making sure requests do not happen twice, is hard.

We could solve that by queueing maybe. 
But would mean we can only do one request at the time, slowing the experience down.

Reply:
"Cancelling requests, while also making sure requests do not happen twice, is hard.". That's because of the tight coupling between components and requests. 

Think it would be best to extract the cache, request queue and logic around clearing the subcaches (groups) and refetching to a global context. Separating the cache completely from the lifetime cycles of the individual components.

Reply:
>> Think it would be best to extract the cache, request queue and logic around clearing the subcaches (groups) and refetching to a global context. Separating the cache completely from the lifetime cycles of the individual components.

How would you that without fetching everything at once at startup?

Reply:
I don't understand the relation you make with lazy fetching vs. initial fetching